### PR TITLE
[SPARK-40688][SQL] Support data masking built-in function 'mask_first_n'

### PR DESCRIPTION
**Pull request by vinodkc**
### What changes were proposed in this pull request?

This PR supports data masking built-in Function **mask_first_n**, 

Returns a masked version of str with the first n values masked. 
By default, first 4 characters will be masked, upper case letters will be converted to "X", lower case letters will be converted to "x" and numbers will be converted to "n".

For example, mask_first_n("**1234-5678-8765-4321**", 4) results in nnnn-5678-8765-4321.
Can override the characters used in the mask by supplying additional arguments: second argument control the number of charatcters to be masked, the third argument controls the mask character for upper case letters, the fourth argument for lower case letters and the fifth argument for numbers.
 
For example, mask_first_n("**abcd-EFGH-8765-4321-abcd-EFGH-8765-4321", 19,  "U", "l", "#**") will return 
**llll-UUUU-####-####-abcd-EFGH-8765-4321**

Examples:
```
  > SELECT _FUNC_('abcd-EFGH-8765-4321');
    xxxx-EFGH-8765-4321
  > SELECT _FUNC_('abcd-EFGH-8765-4321', 9);
    xxxx-XXXX-8765-4321
  > SELECT _FUNC_('abcd-EFGH-8765-@$#', 14);
    xxxx-XXXX-nnnn-@$#
  > SELECT _FUNC_('abcd-EFGH-8765-@$#', 15, 'x', 'X', 'n', 'o');
    XXXXoxxxxonnnno@$#
  > SELECT _FUNC_('abcd-EFGH-8765-@$#', 20, 'x', 'X', 'n', 'o');
    XXXXoxxxxonnnnoooo
  > SELECT _FUNC_('AbCD123-@$#', 10,'Q', 'q', 'd', 'o');
    QqQQdddooo#
  > SELECT _FUNC_('AbCD123-@$#', 10, -1, 'q', 'd', 'o');
    AqCDdddooo#
  > SELECT _FUNC_('AbCD123-@$#', 10, -1, -1, 'd', 'o');
    AbCDdddooo#
  > SELECT _FUNC_('AbCD123-@$#', 10, -1, -1, -1, 'o');
    AbCD123ooo#
  > SELECT _FUNC_(NULL);
    NULL
  > SELECT _FUNC_(NULL, 1, -1, -1, 'o');
    NULL
```

### Why are the changes needed?
To support data masking built-in function **mask_first_n**, which returns a masked version of the input string
Ref : [Data masking functions](https://cwiki.apache.org/confluence/display/hive/languagemanual+udf#LanguageManualUDF-DataMaskingFunctions)

### Does this PR introduce _any_ user-facing change?
Yes, added a new build-in function named '**mask_first_n**' 

### How was this patch tested?
Added test cases

### CODE REVIEW

**Comment by vinodkc** [comment]
@gengliangwang Can you please review this PR ?

**Comment by vinodkc** [comment]
@dtenedor @srielau , Could please review this PR?

**Comment by dtenedor** [comment]
This is very helpful. Thank you @vinodkc for working on this. We will take a look.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala [hunk 1]
```diff
@@ -23,9 +23,54 @@ import org.apache.spark.sql.catalyst.analysis.TypeCheckResult.DataTypeMismatch
 import org.apache.spark.sql.catalyst.expressions.codegen._
 import org.apache.spark.sql.catalyst.expressions.codegen.Block._
 import org.apache.spark.sql.errors.QueryErrorsBase
-import org.apache.spark.sql.types.{AbstractDataType, DataType, StringType}
+import org.apache.spark.sql.types.{AbstractDataType, DataType, IntegerType, StringType}
 import org.apache.spark.unsafe.types.UTF8String
 
+/**
+ * The trait does Input Data Type validation .
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala:30]: please also mention which types of expressions we intend to inherit this trait, perhaps with a couple examples?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala [hunk 2]
```diff
@@ -23,9 +23,54 @@ import org.apache.spark.sql.catalyst.analysis.TypeCheckResult.DataTypeMismatch
...
+      Some(
+        DataTypeMismatch(
+          errorSubClass = "NON_FOLDABLE_INPUT",
+          messageParameters = Map(
+            "inputName" -> message,
+            "inputType" -> toSQLType(exp.dataType),
+            "inputExpr" -> toSQLExpr(exp))))
+    } else {
+      val replaceChar = exp.eval()
+      if (replaceChar != null && replaceChar.asInstanceOf[UTF8String].numChars != 1) {
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala:62]: please use a pattern match against `exp.eval()` instead to (1) check that it is in fact an instance of `UTF8String` and (2) combine the new `if` to the previous `else`, de-denting the rest of the block?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala [hunk 3]
```diff
@@ -257,19 +271,272 @@ case class Mask(
...
+    Arguments:
+      * input      - string value to mask. Supported types: STRING, VARCHAR, CHAR
+      * charCount  - number of characters to be masked. Default value: 4
+      * upperChar  - character to replace upper-case characters with. Specify NULL to retain original character. Default value: 'X'
+      * lowerChar  - character to replace lower-case characters with. Specify NULL to retain original character. Default value: 'x'
+      * digitChar  - character to replace digit characters with. Specify NULL to retain original character. Default value: 'n'
+      * otherChar  - character to replace all other characters with. Specify NULL to retain original character. Default value: NULL
+  """,
+  examples = """
+    Examples:
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala:292]: please also add a case where we request to mask more characters than are present in the input string. What is the behavior?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala [hunk 4]
```diff
@@ -257,19 +271,272 @@ case class Mask(
...
+
+  override def checkInputDataTypes(): TypeCheckResult =
+    validateInputDataTypes(
+      super.checkInputDataTypes(),
+      Seq(
+        (upperChar, "upperChar"),
+        (lowerChar, "lowerChar"),
+        (digitChar, "digitChar"),
+        (otherChar, "otherChar")),
+      () =>
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala:399]: can you please move this lambda into a separate method for better code health?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala [hunk 5]
```diff
@@ -257,19 +271,272 @@ case class Mask(
...
+        Seq(if (!charCountExpr.foldable) {
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala:400]: let's use a pattern match for these checks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala [hunk 6]
```diff
@@ -257,19 +271,272 @@ case class Mask(
...
+
+  /**
+   * Default behavior of evaluation according to the default nullability of QuinaryExpression. If
+   * subclass of QuinaryExpression override nullable, probably should also override this.
+   */
+  override def eval(input: InternalRow): Any = {
+    Mask.mask_first_n(
+      children(0).eval(input),
+      charCount,
+      children(2).eval(input),
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala:438]: we are evaluating all the constant arguments for each row here. Can we please instead perform this once and cache the results instead, for efficiency?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala [hunk 7]
```diff
@@ -257,19 +271,272 @@ case class Mask(
...
-  private val MASKED_LOWERCASE = 'x'
+  val MASKED_LOWERCASE = 'x'
   // Default character to replace digits
-  private val MASKED_DIGIT = 'n'
+  val MASKED_DIGIT = 'n'
   // This value helps to retain original value in the input by ignoring the replacement rules
-  private val MASKED_IGNORE = null
+  val MASKED_IGNORE = null
+  // Default number of characters to be masked
+  val DEFAULT_CHAR_COUNT = 4
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala:537]: please sort these alphabetically

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala (hunk 3 above)
Comment (vinodkc) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala:292]: This case already exists `SELECT _FUNC_('abcd-EFGH-8765-@$#', 20, 'x', 'X', 'n', 'o');`, input string length 18, request to mask 20 chars, 
Result: `XXXXoxxxxonnnnoooo`
Considers all chars  (18)in the input string for masking

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala (hunk 1 above)
Comment (vinodkc) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala]: Moved more common code into this trait from Masking classes  and mentioned additional detail in the description

**Comment by dtenedor** [comment]
@vinodkc please ping again when you're ready for another review.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala (hunk 2 above)
Comment (vinodkc) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala (hunk 4 above)
Comment (vinodkc) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala]: Moved that method to a template pattern

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala (hunk 6 above)
Comment (vinodkc) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala (hunk 7 above)
Comment (vinodkc) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala (hunk 5 above)
Comment (vinodkc) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala]: Done

**Comment by vinodkc** [comment]
@dtenedor , Thank you for the code review comments,  I applied the suggested code changes. Can you please do another review

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala [hunk 8]
```diff
@@ -283,7 +543,32 @@ object Mask {
...
+      maskLower: Any,
+      maskDigit: Any,
+      maskOther: Any): UTF8String = {
+
+    val transformedString = if (input == null) {
+      null
+    } else {
+      val inputStr = input.asInstanceOf[UTF8String]
+      val stringSize = inputStr.numChars
+      val endIdx = if (stringSize < charCount) stringSize else charCount
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala:562]:
```suggestion
      val endIdx = stringSize.min(charCount)
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala [hunk 9]
```diff
@@ -283,7 +543,32 @@ object Mask {
...
+      inputStr.toString.zipWithIndex.map { case (ch, i) =>
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala:563]: `zipWithIndex` is slow, and this expression will be evaluating for every input row. Can we instead split the string based on `endIdx` and then just call `transformChar` on every char of the first token?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala [hunk 10]
```diff
@@ -23,9 +23,82 @@ import org.apache.spark.sql.catalyst.analysis.TypeCheckResult.DataTypeMismatch
 import org.apache.spark.sql.catalyst.expressions.codegen._
 import org.apache.spark.sql.catalyst.expressions.codegen.Block._
 import org.apache.spark.sql.errors.QueryErrorsBase
-import org.apache.spark.sql.types.{AbstractDataType, DataType, StringType}
+import org.apache.spark.sql.types.{AbstractDataType, DataType, IntegerType, StringType}
 import org.apache.spark.unsafe.types.UTF8String
 
+/**
+ * This trait has common methods, which can be used by Classes implementing Masking udf functions
```
[the remark this answers is not in the document]
Comment (vinodkc) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala [hunk 11]
```diff
@@ -23,9 +23,82 @@ import org.apache.spark.sql.catalyst.analysis.TypeCheckResult.DataTypeMismatch
...
+ * Eg : Mask, MaskFirstN, etc
+ */
+private[sql] trait Maskable extends ExpectsInputTypes with QueryErrorsBase {
+  protected val upperCharExpr: Expression
+  protected val lowerCharExpr: Expression
+  protected val digitCharExpr: Expression
+  protected val otherCharExpr: Expression
+
+  @transient
```
[the remark this answers is not in the document]
Comment (vinodkc) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala:44]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala [hunk 12]
```diff
@@ -23,9 +23,82 @@ import org.apache.spark.sql.catalyst.analysis.TypeCheckResult.DataTypeMismatch
...
+  protected lazy val upperChar = upperCharExpr.eval()
+  @transient
+  protected lazy val lowerChar = lowerCharExpr.eval()
+  @transient
+  protected lazy val digitChar = digitCharExpr.eval()
+  @transient
+  protected lazy val otherChar = otherCharExpr.eval()
+
+  protected def validateAdditionalFields(): Seq[Option[TypeCheckResult]] = Seq(None)
```
[the remark this answers is not in the document]
Comment (vinodkc) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala:59]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala [hunk 13]
```diff
@@ -77,13 +150,12 @@ import org.apache.spark.unsafe.types.UTF8String
 // scalastyle:on line.size.limit
 case class Mask(
```
[the remark this answers is not in the document]
Comment (vinodkc) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala:165]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala [hunk 14]
```diff
@@ -236,40 +265,271 @@ case class Mask(
...
+      charCountExpr,
+      upperCharExpr,
+      lowerCharExpr,
+      digitCharExpr,
+      Literal(Mask.MASKED_IGNORE, StringType))
+
+  @transient
+  private lazy val charCount = {
+    val value = charCountExpr.eval().asInstanceOf[Int]
+    if (value < 0) 0 else value
```
[the remark this answers is not in the document]
Comment (vinodkc) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala (hunk 8 above)
Comment (vinodkc) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala (hunk 9 above)
Comment (vinodkc) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala]: Replaced `zipWithIndex` with `for i`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala [hunk 15]
```diff
@@ -236,40 +265,271 @@ case class Mask(
...
+      digitCharExpr = newChildren(3),
+      otherCharExpr = newChildren(4))
+}
+
+// scalastyle:off line.size.limit
+@ExpressionDescription(
+  usage =
+    """_FUNC_(input[, charCount, upperChar, lowerChar, digitChar, otherChar]) - masks the first n characters of given string value.
+       The function masks the first n characters of the value with 'X' or 'x', and numbers with 'n'.
+       This can be useful for creating copies of tables with sensitive information removed.
```
[the remark this answers is not in the document]
Comment (vinodkc) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala:303]: Done

**Review by dtenedor — APPROVED** [review]
This looks good. Thanks again for doing the work to implement this functionality!

**Comment by dtenedor** [comment]
@gengliangwang do you want to take a look and possibly help with merging this once we feel all the reviewers have approved?

**Comment by vinodkc** [comment]
@gengliangwang Could please review this PR?

**Comment by dtenedor** [comment]
@cloud-fan would you mind helping review this as well? It LGTM as of now.

**Comment by vinodkc** [comment]
@cloud-fan Would please help review this PR?

**Pull request closed without merging**